File: backend/app/schemas/runner_status.py
"""Pydantic schemas for runtime executor / runner observability."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class RunnerOperationRecordResponse(BaseModel):
    operation: str
    provider: str
    status: str
    duration_ms: int
    request_id: str | None = None
    deployment_id: str | None = None
    topology_id: str | None = None
    error_message: str | None = None
    created_at: datetime | str


class RecentRunnerOperationsResponse(BaseModel):
    operations: list[RunnerOperationRecordResponse] = Field(default_factory=list)
    count: int = 0


class RunnerStatusDetailResponse(BaseModel):
    runner_reachable: bool
    runtime_executor: str
    runner_status: str | None = None
    status: str | None = None
    runtime_provider: str | None = None
    docker_reachable: bool | None = None
    kubernetes_reachable: bool | None = None
    current_context: str | None = None
    version: str | None = None
    git_sha: str | None = None
    build_time: str | None = None
    supported_operations: list[str] = Field(default_factory=list)
    last_runtime_error: str | None = None
    message: str | None = None
    checked_at: datetime | str

File: frontend/src/types/runnerStatus.ts
export type LastRuntimeErrorDetail = {
  operation: string;
  message: string;
  request_id?: string | null;
  status_code?: number | null;
  timestamp: string;
  historical?: boolean;
};

export type RunnerStatusDetail = {
  runner_reachable: boolean;
  runtime_executor: string;
  runner_status?: string | null;
  status?: string | null;
  runtime_provider?: string | null;
  docker_reachable?: boolean | null;
  kubernetes_reachable?: boolean | null;
  current_context?: string | null;
  version?: string | null;
  git_sha?: string | null;
  build_time?: string | null;
  supported_operations?: string[];
  last_runtime_error?: LastRuntimeErrorDetail | null;
  message?: string | null;
  checked_at?: string;
};

export type RunnerOperationRecord = {
  operation: string;
  provider: string;
  status: string;
  duration_ms: number;
  request_id?: string | null;
  deployment_id?: string | null;
  topology_id?: string | null;
  error_message?: string | null;
  status_code?: number | null;
  created_at: string;
};

export type RecentRunnerOperationsResponse = {
  operations: RunnerOperationRecord[];
  count: number;
};

export type RuntimeStatusResponse = {
  backend_status?: string;
  status?: string;
  runtime_executor?: string;
  runtime_provider?: string;
  runner_reachable?: boolean;
  docker_reachable?: boolean;
  kubernetes_reachable?: boolean;
  current_context?: string;
  kubeconfig_source?: string;
  kubernetes_init_error?: string;
  message?: string;
  last_runtime_error?: LastRuntimeErrorDetail | null;
  environment?: string;
  checked_at?: string;
  runner?: RunnerStatusDetail;
  supported_operations?: string[];
  version?: string;
  git_sha?: string;
  build_time?: string;
};

export function formatLastRuntimeError(detail: LastRuntimeErrorDetail | null | undefined): string | null {
  if (!detail?.operation) return null;
  const parts = [`Last failed operation: ${detail.operation}`];
  if (detail.status_code != null) {
    parts.push(`returned ${detail.status_code}`);
  }
  if (detail.message) {
    parts.push(`— ${detail.message}`);
  }
  if (detail.request_id) {
    parts.push(`— request_id ${detail.request_id}`);
  }
  if (detail.timestamp) {
    parts.push(`(${new Date(detail.timestamp).toLocaleString()})`);
  }
  if (detail.historical) {
    parts.push('(historical)');
  }
  return parts.join(' ');
}

export function pickActiveRuntimeError(
  runnerStatus: RunnerStatusDetail | null | undefined,
  runtimeStatus: RuntimeStatusResponse | null | undefined,
): LastRuntimeErrorDetail | null {
  const candidates = [
    runnerStatus?.last_runtime_error,
    runtimeStatus?.last_runtime_error,
    runtimeStatus?.runner?.last_runtime_error,
  ];
  for (const err of candidates) {
    if (err && !err.historical) {
      return err;
    }
  }
  return null;
}
